scanner/decisions/0009-nvd-cvss-scores.md

```diff
@@ -0,0 +1,121 @@
+
+## Status
+
+Status: Accepted.
+
+## Context
+
+Users who require FedRAMP compliance [must be able see NVD's CVSS scores associated with each CVE](https://www.fedramp.gov/assets/resources/documents/CSP_Vulnerability_Scanning_Requirements.pdf).
+
+Currently, Scanner V4 provides a single CVSS score and severity, which is preferably from the vendor.
+In reality, the CVSS scores that are shown are from NVD except in the following cases:
+
+* RHEL-base images
+  * Red Hat-provided CVSS scores/severities are used.
+* OSV-provided data
+  * If OSV.dev has a CVSS score, then that score is used.
+    * The score is converted into a severity based on the CVSS version's specification.
+  * Note: it is likely, if not always true, the score displayed by OSV is the same as NVD.
+
+The current API looks like the following:
+
+```
+message VulnerabilityReport {
+  message Vulnerability {
+    enum Severity {
+      SEVERITY_UNSPECIFIED = 0;
+      SEVERITY_LOW = 1;
+      SEVERITY_MODERATE = 2;
+      SEVERITY_IMPORTANT = 3;
+      SEVERITY_CRITICAL = 4;
+    }
+    message CVSS {
+      message V2 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      message V3 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      V2 v2 = 1;
+      V3 v3 = 2;
+    }
+    ...
+    string severity = 6;
+    ...
+    CVSS cvss = 12;
+  }
+  ...
+}
+```
+
+## Decision
+
+The VulnerabilityReport API will be extended to support more than one CVSS message. This will be done as follows:
+
+```
+message VulnerabilityReport {
+  message Vulnerability {
+    enum Severity {
+      SEVERITY_UNSPECIFIED = 0;
+      SEVERITY_LOW = 1;
+      SEVERITY_MODERATE = 2;
+      SEVERITY_IMPORTANT = 3;
+      SEVERITY_CRITICAL = 4;
+    }
+    message CVSS {
+      message V2 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      message V3 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      V2 v2 = 1;
+      V3 v3 = 2;
+    }
+    ...
+    string severity = 6;
+    ...
+    CVSS cvss = 12;
+    CVSS nvd_cvss = 13; <-- New field.
+  }
+  ...
+}
+```
+
+Currently, there is only a need to have CVSS scores from the vendor (if available) and NVD (if available), no more.
+Making this explicit, then, makes it easy and clear to users to find each scores and choose between the two.
+
+Another option is to provide a more generic solution. One way to do this would be to update `CVSS cvss = 12` to 
+`repeated CVSS cvss = 12` (note: the [proto3 spec](https://protobuf.dev/programming-guides/proto3/#updating) indicates this is compatible).
+Doing it this way gives us a more generic way of supporting multiple CVSS scores vs adding a new `nvd_cvss` field,
+and it may certainly be the way to go in the future should we want to support a variety of score sources; however,
+there is no need for a generic solution at this time, and adding the new field make it very easy to choose between the two options.
```

Comment:
The generic solution would benefit from being  a 'map' of sorts instead of a array (conceptually, ignoring the wire format), the value would have to be generic enough to support serialization/deserialization to different types (like what is done in the storage protos today).

Making this generic in the future would likely require a new field as a result, and then modifying logic to prefer the 'scores from the map' if they exist, otherwise the static fields - so I don't believe explicit solution hurts the move to generic in the future if needed.

Reply:
Yes, I agree. I think it would require some thought to determine a nice namespace of the "keys" (for example, domain names could make sense like "redhat.com" or "nvd.nist.org") for it to be fully decoupled.